File: cpp/queen-attack/queen_attack.cpp
#include "queen_attack.h"

#include <stdexcept>

namespace queen_attack {

chess_board::chess_board(const coord &white, const coord &black) {
  if (!valid_position(white) || !valid_position(black)) {
    throw std::domain_error(
        "Invalid Position: Queens are outside the board boundaries.");
  }
  if (white == black) {
    throw std::domain_error(
        "Invalid Position: The two queens cannot occupy the same space.");
  }
  this->_white = white;
  this->_black = black;
}

coord chess_board::white() const { return this->_white; }

coord chess_board::black() const { return this->_black; }

bool chess_board::can_attack() const {
  return this->same_column() || this->same_row() || this->same_diagonal();
}

bool chess_board::same_column() const {
  return this->_white.first == this->_black.first;
}

bool chess_board::same_row() const {
  return this->_white.second == this->_black.second;
}

bool chess_board::same_diagonal() const {
  return this->delta_column() == this->delta_row();
}

bool chess_board::valid_position(const coord &position) {
  if (position.first < BOARD_START || position.first > BOARD_END) {
    return false;
  }
  if (position.second < BOARD_START || position.second > BOARD_END) {
    return false;
  }
  return true;
}

int chess_board::delta_column() const {
  return std::abs(this->_white.first - this->_black.first);
}

int chess_board::delta_row() const {
  return std::abs(this->_white.second - this->_black.second);
}

}  // namespace queen_attack

File: go/queen-attack/queen_attack.go
// Package queenattack provides a function to determine if two queens on a standard chessboard
// can attack each other.
package queenattack

import (
	"errors"
	"regexp"
)

// CanQueenAttack determines if two queens can attack each other on a chessboard.
// The positions are given in algebraic notation, e.g., "a1", "h8".
// It returns true if an attack is possible, along with a nil error.
// It returns false and an error if the positions are invalid or identical.
func CanQueenAttack(whitePosition, blackPosition string) (bool, error) {
	whiteOk, whiteRow, whiteCol := parse(whitePosition)
	blackOk, blackRow, blackCol := parse(blackPosition)

	if !whiteOk || !blackOk {
		return false, errors.New("invalid position on the board")
	}

	deltaRow := abs(blackRow - whiteRow)
	deltaCol := abs(blackCol - whiteCol)

	if deltaCol == 0 && deltaRow == 0 {
		return false, errors.New("queens cannot occupy the same square")
	}

	return deltaRow == 0 || deltaCol == 0 || deltaRow == deltaCol, nil
}

// parse converts a chessboard position string (e.g., "a1") into a row and column integer.
// It returns a boolean indicating if the parsing was successful.
func parse(position string) (bool, int, int) {
	if regexp.MustCompile("^[a-h][1-8]$").MatchString(position) {
		runes := []rune(position)

		row := int(runes[0] - 'a')
		col := int(runes[1] - '0')

		return true, row, col
	}
	return false, 0, 0
}

// abs returns the absolute value of an integer.
func abs(value int) int {
	if value < 0 {
		return -value
	}
	return value
}
